### DISCUSSION TITLE
Preserving ExcludeFromCodeCoverage  attribute

### DISCUSSION THREAD

**Discussion by jakubch1** [dotnet/roslyn#49663] [Q&A]
Hello,
Currently ExcludeFromCodeCoverage  attribute is not preserved when roslyn generates additional methods/classes (async methods, lambda... etc). To fix it code coverage tools needs to add logic which depends on internal roslyn implementation:

```
        // Check if the name is synthesized by the compiler
        // Refer to https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Symbols/Synthesized/GeneratedNames.cs
        // to see how the compiler generate names for lambda, local function, yield or async/await expressions
        internal bool IsSynthesizedNameOf(string name, string methodName, int methodOrdinal)
        {
            return
                // Lambda method
                name.IndexOf($"<{methodName}>b__{methodOrdinal}") != -1 ||
                // Lambda display class
                name.IndexOf($"<>c__DisplayClass{methodOrdinal}_") != -1 ||
                // State machine
                name.IndexOf($"<{methodName}>d__{methodOrdinal}") != -1 ||
                // Local function
                (name.IndexOf($"<{methodName}>g__") != -1 && name.IndexOf($"|{methodOrdinal}_") != -1);
        }
```
[from https://github.com/coverlet-coverage/coverlet/pull/671/files]

Is there any better way of fixing it?

For dynamic code coverage we received also several issues. For example:
https://github.com/dotnet/core/issues/4361
https://github.com/dotnet/core/issues/5664

We are wondering if we should create similar logic as coverlet.

BR,
Jakub

**Comment by tmat** [comment]
@jaredpar @jcouv

**Comment by tmat** [comment]
One option would be to add an attribute that can be applied to attributes and indicate that specified attribute should be emitted to all members generated for the annotated member. Another option would be to propagate all attributes - but that might not be desirable (e.g. `[Fact]` attribute probably shouldn't be).

**Comment by jakubch1** [comment]
@jaredpar @tmat  @jcouv any update on this?
